File: leetcode-cn/2511.0_Maximum_Enemy_Forts_That_Can_Be_Captured.py
'''
two pointers
T: O(N)
S: O(1)

48ms 击败 33.88%使用 Python3 的用户
15.78MB 击败 15.70%使用 Python3 的用户
'''
class Solution:
    def captureForts(self, forts: List[int]) -> int:
        ans = i = j = 0
        n = len(forts)
        while i < n:
            while i < n and forts[i] == 0:
                i += 1
            j = i + 1
            while j < n and forts[j] == 0:
                j += 1
            if j < n and forts[i] + forts[j] == 0:
                ans = max(ans, j - i - 1)
            i = j
        return ans 


'''
40ms 击败 81.82%使用 Python3 的用户
15.74MB 击败 26.45%使用 Python3 的用户
'''
class Solution:
    def captureForts(self, forts: List[int]) -> int:
        ans, pre = 0, -1
        for i, fort in enumerate(forts):
            if fort != 0:
                if pre >= 0 and fort != forts[pre]:
                    ans = max(ans, i - pre - 1)
                pre = i 
        return ans 

File: js/2511.0_Maximum_Enemy_Forts_That_Can_Be_Captured.js
/*
60ms, 击败 66.67%使用 JavaScript 的用户
39.81MB, 击败 94.87%使用 JavaScript 的用户
*/
/**
 * @param {number[]} forts
 * @return {number}
 */
var captureForts = function(forts) {
    let ans = 0, pre = -1;
    for (let i = 0; i < forts.length; i++) {
        if (forts[i] != 0) {
            if (pre >= 0 && forts[i] != forts[pre]) {
                ans = Math.max(ans, i - pre - 1);
            }
            pre = i;
        }
    }
    return ans;
};


/*
var, not let
64ms, 击败 41.03%使用 JavaScript 的用户
40.02MB, 击败 76.92%使用 JavaScript 的用户
*/
/**
 * @param {number[]} forts
 * @return {number}
 */
var captureForts = function(forts) {
    var ans = 0, pre = -1;
    for (var i = 0; i < forts.length; i++) {
        if (forts[i] != 0) {
            if (pre >= 0 && forts[i] != forts[pre]) {
                ans = Math.max(ans, i - pre - 1);
            }
            pre = i;
        }
    }
    return ans;
};
